ipa-core/src/protocol/ipa_prf/prf_sharding/mod.rs

```diff
@@ -401,23 +397,35 @@ where
 /// # Panics
 /// Propagates errors from multiplications
 #[tracing::instrument(name = "attribute_cap_aggregate", skip_all)]
-pub async fn attribute_cap_aggregate<'ctx, BK, TV, HV, TS, const SS_BITS: usize, const B: usize>(
-    sh_ctx: SemiHonestContext<'ctx>,
+pub async fn attribute_cap_aggregate<
+    'ctx,
+    C,
+    BK,
+    TV,
+    HV,
+    TS,
+    const SS_BITS: usize,
+    const B: usize,
+>(
+    sh_ctx: C,
     input_rows: Vec<PrfShardedIpaInputRow<BK, TV, TS>>,
     attribution_window_seconds: Option<NonZeroU32>,
     histogram: &[usize],
 ) -> Result<BitDecomposed<Replicated<Boolean, B>>, Error>
 where
+    C: UpgradableContext + 'ctx,
     BK: BreakdownKey<B>,
     TV: BooleanArray + U128Conversions,
     HV: BooleanArray + U128Conversions,
     TS: BooleanArray + U128Conversions,
     Boolean: FieldSimd<B>,
-    Replicated<Boolean, B>:
-        BooleanProtocols<UpgradedSemiHonestContext<'ctx, NotSharded, Boolean>, B>,
-    for<'a> Replicated<BK>: BooleanArrayMul<UpgradedSemiHonestContext<'a, NotSharded, Boolean>>,
-    for<'a> Replicated<TS>: BooleanArrayMul<UpgradedSemiHonestContext<'a, NotSharded, Boolean>>,
-    for<'a> Replicated<TV>: BooleanArrayMul<UpgradedSemiHonestContext<'a, NotSharded, Boolean>>,
+    Replicated<Boolean>: BooleanProtocols<<C::DZKPValidator as DZKPValidator>::Context>,
```

Comment:
I added a type alias so we can say `BooleanProtocols<DZKPUpgraded<C>>` instead of `BooleanProtocols<<C::DZKPValidator as DZKPValidator>::Context>`. I'm not sure I understand the `validate_record` suggestion.

Reply:
I was thinking that `validate_record` can be used on the `Context` type, but after I implemented `prf_eval` with it, I still had to plumb `V as Validator>:UpgradedContext` everywhere, so this suggestion won't help